### PR TITLE
KT-30643 J2K: wrong position of TYPE_USE annotation

### PR DESCRIPTION
**Pull request by t-kameyama**
https://youtrack.jetbrains.com/issue/KT-30643

### CODE REVIEW

File: nj2k/src/org/jetbrains/kotlin/nj2k/JavaToJKTreeBuilder.kt [hunk 1]
```diff
@@ -719,11 +717,12 @@ class JavaToJKTreeBuilder constructor(
             visibility(referenceSearcher) { ast, psi -> ast.withFormattingFrom(psi) }
 
         fun PsiField.toJK(): JKField {
+            val (annotationList, typeUseAnnotationList) = annotationList(this)
             return JKField(
-                JKTypeElement(type.toJK()).withFormattingFrom(typeElement),
+                JKTypeElement(type.toJK(), typeUseAnnotationList).withFormattingFrom(typeElement),
```
Comment (darthorimar) [inline nj2k/src/org/jetbrains/kotlin/nj2k/JavaToJKTreeBuilder.kt]: Please, take annotations directly from `PsiField.typeElement`

File: nj2k/src/org/jetbrains/kotlin/nj2k/JavaToJKTreeBuilder.kt [hunk 2]
```diff
@@ -735,18 +734,29 @@ class JavaToJKTreeBuilder constructor(
             }
         }
 
-        fun <T : PsiModifierListOwner> T.annotationList(docCommentOwner: PsiDocCommentOwner?): JKAnnotationList {
+        fun <T : PsiModifierListOwner> T.annotationList(docCommentOwner: PsiDocCommentOwner?): Pair<JKAnnotationList, JKAnnotationList> {
```
Comment (darthorimar) [inline nj2k/src/org/jetbrains/kotlin/nj2k/JavaToJKTreeBuilder.kt]: And in this function just filter out type-use annotations by checking `owner` property

File: nj2k/src/org/jetbrains/kotlin/nj2k/JavaToJKTreeBuilder.kt [hunk 3]
```diff
@@ -735,18 +736,23 @@ class JavaToJKTreeBuilder constructor(
...
+        fun <T : PsiModifierListOwner> T.annotationList(
+            docCommentOwner: PsiDocCommentOwner?,
+            typeAnnotations: Array<out PsiAnnotation>
+        ): JKAnnotationList {
             val deprecatedAnnotation = docCommentOwner?.docComment?.deprecatedAnnotation()
             val plainAnnotations = annotations.mapNotNull { annotation ->
                 when {
                     annotation !is PsiAnnotation -> null
                     annotation.qualifiedName == DEPRECATED_ANNOTAION_FQ_NAME && deprecatedAnnotation != null -> null
+                    annotation in typeAnnotations -> null
```
Comment (t-kameyama) [inline nj2k/src/org/jetbrains/kotlin/nj2k/JavaToJKTreeBuilder.kt]: `it.owner` returns PsiModifierList, not PsiTypeElement. Could you tell me how to filter out by `it.owner`?

Comment (darthorimar) [reply nj2k/src/org/jetbrains/kotlin/nj2k/JavaToJKTreeBuilder.kt]: As far, as I can see `com.intellij.psi.PsiAnnotation#getOwner` has type `PsiAnnotationOwner` and `PsiTypeElement` is subtype of it

File: nj2k/src/org/jetbrains/kotlin/nj2k/JavaToJKTreeBuilder.kt [hunk 4]
```diff
@@ -810,16 +816,19 @@ class JavaToJKTreeBuilder constructor(
...
         fun PsiMethod.toJK(): JKMethod {
+            val returnType = returnType
+            val returnTypeAnnotations = returnType?.annotations.orEmpty()
             return JKMethodImpl(
                 JKTypeElement(
                     returnType?.toJK()
                         ?: JKJavaVoidType.takeIf { isConstructor }
-                        ?: JKNoType),
+                        ?: JKNoType,
+                    returnTypeAnnotations.toJK()),
```
Comment (darthorimar) [inline nj2k/src/org/jetbrains/kotlin/nj2k/JavaToJKTreeBuilder.kt]: Let's create `PsiTypeElement.toJk()` which will also retrieve annotations from `PsiTypeElement`

File: nj2k/src/org/jetbrains/kotlin/nj2k/JavaToJKTreeBuilder.kt (hunk 3 above)
Comment (t-kameyama) [reply nj2k/src/org/jetbrains/kotlin/nj2k/JavaToJKTreeBuilder.kt]: I think `getOwner()` returns `PsiModifierList` here:
https://github.com/JetBrains/intellij-community/blob/master/java/java-psi-impl/src/com/intellij/psi/impl/source/tree/java/PsiAnnotationImpl.java#L120-L124

Comment (darthorimar) [reply nj2k/src/org/jetbrains/kotlin/nj2k/JavaToJKTreeBuilder.kt]: You are right :) then `com.intellij.codeInsight.AnnotationTargetUtil#isTypeAnnotation` is your friend here

File: nj2k/src/org/jetbrains/kotlin/nj2k/JavaToJKTreeBuilder.kt [hunk 5]
```diff
@@ -719,8 +719,9 @@ class JavaToJKTreeBuilder constructor(
             visibility(referenceSearcher) { ast, psi -> ast.withFormattingFrom(psi) }
 
         fun PsiField.toJK(): JKField {
+            val typeElement = typeElement
```
Comment (darthorimar) [inline nj2k/src/org/jetbrains/kotlin/nj2k/JavaToJKTreeBuilder.kt]: Why we need this variable?

File: nj2k/src/org/jetbrains/kotlin/nj2k/JavaToJKTreeBuilder.kt [hunk 6]
```diff
@@ -862,9 +868,10 @@ class JavaToJKTreeBuilder constructor(
                 }
 
 
-        fun PsiLocalVariable.toJK(): JKLocalVariable =
-            JKLocalVariable(
-                JKTypeElement(type.toJK()).withFormattingFrom(typeElement),
+        fun PsiLocalVariable.toJK(): JKLocalVariable {
+            val typeElement = typeElement
```
Comment (darthorimar) [inline nj2k/src/org/jetbrains/kotlin/nj2k/JavaToJKTreeBuilder.kt]: Redundant var?

File: nj2k/src/org/jetbrains/kotlin/nj2k/JavaToJKTreeBuilder.kt (hunk 5 above)
Comment (t-kameyama) [reply nj2k/src/org/jetbrains/kotlin/nj2k/JavaToJKTreeBuilder.kt]: `this.typeElement` calls: 
https://github.com/JetBrains/intellij-community/blob/master/java/java-psi-impl/src/com/intellij/psi/impl/source/PsiFieldImpl.java#L132-L139

We should avoid unnecessary method calls.
https://github.com/t-kameyama/kotlin/blob/07054bdee360b173175d4df6ade67a209ad504cd/nj2k/src/org/jetbrains/kotlin/nj2k/JavaToJKTreeBuilder.kt#L724

File: nj2k/src/org/jetbrains/kotlin/nj2k/JavaToJKTreeBuilder.kt (hunk 6 above)
Comment (t-kameyama) [reply nj2k/src/org/jetbrains/kotlin/nj2k/JavaToJKTreeBuilder.kt]: Same as above.

File: nj2k/src/org/jetbrains/kotlin/nj2k/JavaToJKTreeBuilder.kt (hunk 5 above)
Comment (darthorimar) [reply nj2k/src/org/jetbrains/kotlin/nj2k/JavaToJKTreeBuilder.kt]: This call is very cheap, so there is no need for that :)

But for avoiding this you can move all functionality related to working with type elements to separated function like `PsiTypeElement.toJk(): JKtypeElement` which will retrieve type, annotations, and formating from type element

**Comment by darthorimar** [comment]
@t-kameyama, two tests are failing now:

1. `NewJavaToKotlinConverterSingleFileTestGenerated.ArrayInitializerExpression#testDoubleArrayWithVariables`
2. `NewJavaToKotlinConverterSingleFileTestGenerated.ArrayInitializerExpression#testOneDimWithVariables`

Please, consider fixing them

**Comment by darthorimar** [comment]
@t-kameyama thanks!

**Pull request merged**